Project Manager Native Image

radeusgd opened the pull request:
### Pull Request Description

<!--
- Please describe the nature of your PR here, as well as the motivation for it.
- If it fixes an open issue, please mention that issue number here.
-->

Closes #1150 

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

- Marked as breaking because `bin/project-manager.bat` is now `bin/project-manager.exe` in the Windows distribution.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

4e6 commented on .gitignore:
```diff
@@ -83,11 +83,15 @@ bench-report.xml
 *.jar
+/built-engine/
+/built-project-manager/
+/built-launcher/
```
Instead of creating 4 separate directories in the root, I'd create a single one (with subdirectories) to store the build output. IMO this way it makes less noise when you try to `ls` things

4e6 commented on .gitignore:
```diff
@@ -83,11 +83,15 @@ bench-report.xml
 ##############
 
 /enso
+/project-manager
```
Why do we need `project-manager` directory in addition to the `built-project-manager`?

radeusgd replied:
`project-manager` is not a directory, it's the name of the executable when building the project manager on UNIX platforms.

radeusgd replied:
And `built-project-manager` actually represents a built distribution package (maybe I should rename it to `distribution-packages/project-manager`; while the `project-manager` executable is just an executable that we build, similarly as `runner.jar` etc. that we keep in the root.

kustosz reviewed (commented):
I'm unable to review this. It's 10k lines of random stuff – licenses, mixed with code files, mixed with some configs (are these autogenerated? are these manually created?). There is no way I can understand it in a finite time.

radeusgd commented:
> I'm unable to review this. It's 10k lines of random stuff – licenses, mixed with code files, mixed with some configs (are these autogenerated? are these manually created?). There is no way I can understand it in a finite time.

I was considering splitting out the license review into a separate PR but wanted to avoid having a state where we technically distribute the new and separate PM without its licenses; in hindisght I should have made a separate PR for native-image and separate for splitting distribution.

I think you can rather safely skip the license-related files, especially the ones in `THIRD-PARTY` dir as they are auto-generated. Our license tool has accepted this review and the way I created the license review for PM was just by copying the relevant parts from the engine distribution - so nothing new here, just shuffling stuff around to accomodate for the split of the distributions.

I think probably using this filter https://github.com/enso-org/enso/pull/1318/files?file-filters%5B%5D=.bat&file-filters%5B%5D=.conf&file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.properties&file-filters%5B%5D=.sbt&file-filters%5B%5D=.scala&file-filters%5B%5D=.sh&file-filters%5B%5D=.yml&file-filters%5B%5D=dotfile should show you mostly relevant stuff.

You can also mostly ignore `*-config.json`, they were generated semi-automatically, according to the process described in [our docs](https://github.com/enso-org/enso/blob/wip/rw/pm-native-image/docs/infrastructure/native-image.md#Configuration).

Summary of what has changed:
- parts of Native Image config were migrated to `native-image.properties` in `META-INF` and instead of passing a classpath we build a JAR and pass path to this JAR - this was to reduce the commandline length,
- in the CI, we now use SBT tasks to build the packages, they were added in file `project/DistributionPackage.scala` and `build.sbt`,
- added a tool for cleaning up the autogenerated reflection configs, as described in doc linked above,
- the PM is now a separate artifact on CI,
- minor fixes for logging in the launcher,
- updated docs to reflect that PM is no longer distributed with engine.

kustosz commented on .github/workflows/release.yml:
```diff
@@ -498,6 +478,40 @@ jobs:
           asset_name: enso-launcher-${{ env.DIST_VERSION }}-windows-amd64.zip
           asset_content_type: application/zip
 
+      - name: Publish the Project Manager (Linux)
```
Shouldn't these check what platform is being run?

radeusgd replied:
No, this is a separate job that is composing the packages and it is only running on a Linux runner - it uses artifacts built in an earlier stage on different platforms and packages them.

kustosz commented on .gitignore:
```diff
@@ -83,11 +83,13 @@ bench-report.xml
 /enso.lib
 *.pdb
 *.jar
+/built-distribution/
```
Why not just drop the artifacts into `distribution`, like we do now?

radeusgd replied:
The contents of `built-distribution` already have a layout that is ready to be packaged, while `distribution` is currently a bit messy - there is a separate `engine` directory but parts of the `distribution` root directory also belong in `engine` package (like the `std-lib` directory).

I think what you suggested is a good idea but it requires changing the structure of `distribution` directory. I think maybe it should be done as part of #1046?

kustosz reviewed (approved):
It's OK, I guess
